### PR TITLE
Maltiverse Integration

### PR DESCRIPTION
**Pull request by darkushin**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
New integration for Maltiverse reputation

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [x] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [X] Tests
- [X] Documentation 

## Demisto Partner?
- [ ] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.yml [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+  feed: false
+  isfetch: false
+  longRunning: false
+  longRunningPort: false
+  runonce: false
+  script: '-'
+  subtype: python3
+  type: python
+tests:
+- No test
```
Comment (yuvalbenshalom) [inline Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.yml:415]: the test playbook should be here

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse_description.md [hunk 2]
```diff
@@ -0,0 +1 @@
+## Add description
```
Comment (yuvalbenshalom) [inline Packs/Maltiverse/Integrations/Maltiverse/Maltiverse_description.md:1]: here we should have an explanation about where to get the API key from

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+        dbot_score = {'Indicator': report.get('ip_addr', ''), 'Type': 'ip', 'Vendor': 'Maltiverse',
+                      'Score': calculate_score(positive_detections, report.get('classification', ''), threshold)}
+
+        maltiverse_ip = {**blacklist_context, **additional_info}
+
+        context[outputPaths['ip']].append(outputs)
+        context[f'Maltiverse.{outputPaths["ip"]}'].append(maltiverse_ip)
+        context[DBOT_SCORE_KEY].append(dbot_score)
+
+        markdown += tableToMarkdown(f'Maltiverse IP reputation for: {report["ip_addr"]}\n', outputs, removeNull=True)
```
Comment (yuvalbenshalom) [inline Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py:176]:
```suggestion
        markdown += tableToMarkdown(f'Maltiverse IP reputation for: {report['ip_addr']}\n', outputs, removeNull=True)
```

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+        return 0
+    elif classification == 'whitelist':
+        return 1
+    elif positive_detections <= threshold and classification != 'malicious':
+        if anti_virus > 1:
+            return 3
+        return 2
+    elif positive_detections > threshold or classification == 'malicious':
+        return 3
+    else:  # if reached this line there is a problem with the logic
```
Comment (amshamah419) [inline Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py:86]: Is there a more graceful way of handling this?

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+                    'Vendor': 'Maltiverse'
+                }
+            }
+            outputs = {**outputs, **malicious_info}
+            md_info['URL.Malicious.Description'] = blacklist_context['Blacklist']['Description']
+            md_info['URL.Malicious.Vendor'] = 'Maltiverse'
+
+        context[outputPaths['url']].append(outputs)
+        context[DBOT_SCORE_KEY].append(dbot_score)
+        context[f'Maltiverse.{outputPaths["url"]}'].append(maltiverse_url)
```
Comment (yuvalbenshalom) [inline Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py:237]:
```suggestion
        context[f'Maltiverse.{outputPaths['url']}'].append(maltiverse_url)
```

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+
+        markdown += tableToMarkdown(f'Maltiverse URL Reputation for: {report["url"]}\n',
```
Comment (yuvalbenshalom) [inline Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py:239]:
```suggestion
        markdown += tableToMarkdown(f'Maltiverse URL Reputation for: {report['url']}\n',
```

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+        maltiverse_domain = {string_to_context_key(field): report.get(field, '') for field in
+                             ['creation_time', 'modification_time', 'tld', 'classification', 'tag']
+                             }
+        maltiverse_domain['Address'] = report.get('hostname', '')
+        maltiverse_domain = {**maltiverse_domain, **blacklist_context}
+        maltiverse_domain = {**maltiverse_domain, **resolvedIP_info}
+
+        context[outputPaths['domain']].append(outputs)
+        context[DBOT_SCORE_KEY].append(dbot_score)
+        context[f'Maltiverse.{outputPaths["domain"]}'].append(maltiverse_domain)
```
Comment (yuvalbenshalom) [inline Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py:295]:
```suggestion
        context[f'Maltiverse.{outputPaths['domain']}'].append(maltiverse_domain)
```

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+            'Domain.Name': report.get('hostname', ''),
+            'Domain.CreationDate': report.get('creation_time', ''),
+            'Domain.ModificationDate': report.get('modification_time', ''),
+            'Maltiverse.Domain.ModificationTime': report.get('modification_time', ''),
+            'Maltiverse.Domain.CreationTime': report.get('creation_time', ''),
+            'Maltiverse.Domain.ResolvedIP.IP': [report['resolved_ip'][i]['ip_addr'] for i in
+                                                range(len(report['resolved_ip']))]
+        }
+
+        markdown += tableToMarkdown(f'Maltiverse Domain Reputation for: {report["hostname"]}\n',
```
Comment (yuvalbenshalom) [inline Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py:307]:
```suggestion
        markdown += tableToMarkdown(f'Maltiverse Domain Reputation for: {report['hostname']}\n',
```

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.yml [hunk 9]
```diff
@@ -0,0 +1,415 @@
+category: Data Enrichment & Threat Intelligence
+commonfields:
+  id: Maltiverse
+  version: -1
+configuration:
+- display: API Key
+  name: api_key
+  required: false
+  type: 12
```
Comment (amshamah419) [inline Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.yml:9]:
```suggestion
  type: 4
```
Generally, we want to keep API keys secret.

File: Packs/Maltiverse/README.md [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+After you successfully execute a command, a DBot message appears in the War Room with the command details.
+1. ip
+2. domain
+3. url
+4. file
+### 1. ip
+---
+Checks the reputation of an IP address
+##### Required Permissions
+**FILL IN REQUIRED PERMISSIONS HERE**
```
Comment (amshamah419) [inline Packs/Maltiverse/README.md:38]: If the API doesn't require specific permissions at all, we can go ahead and delete this.

File: Packs/Maltiverse/README.md [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+| Maltiverse.Domain.ResolvedIP.IP | String | Stores an IP that was resolved by the domain | 
+| Maltiverse.Domain.ResolvedIP.Timestamp | Date | Stores an timestamp when an IP address has been resolved by the domain | 
+
+
+##### Command Example
+```!domain domain=google.com```
+
+##### Human Readable Output
+
+
```
Comment (amshamah419) [inline Packs/Maltiverse/README.md:127]: Need to double-check, but historically we would do a screenshot of what get's returned to the war room.

File: Packs/Maltiverse/pack_metadata.json [hunk 12]
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "Maltiverse",
+  "description": "Maltiverse helps you to analyze suspicious hashes, URLs, domains, and IP addresses.\n",
```
Comment (amshamah419) [inline Packs/Maltiverse/pack_metadata.json:3]: Not sure we need the `\n` here

File: Packs/Maltiverse/README.md [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+---
+
+## Troubleshooting
+---
+
+
+## Possible Errors (DO NOT PUBLISH ON ZENDESK):
+* 'The given IP was invalid'
+* 'Command not found.'
+* f'Failed to execute {command} command. Error: {e}'
```
Comment (amshamah419) [inline Packs/Maltiverse/README.md:248]: Make sure to fill out what you can here.

File: Packs/Maltiverse/README.md [hunk 14]
```diff
@@ -0,0 +1,248 @@
+## Overview
+---
+
+Analyze suspicious hashes, URLs, domains and IP addresses
+This integration was integrated and tested with version xx of Maltiverse
```
Comment (amshamah419) [inline Packs/Maltiverse/README.md:5]: Is there an API version? If so, go ahead and change the `xx` to the version.

File: Packs/Maltiverse/README.md [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+## Maltiverse Playbook
+---
+
```
Comment (amshamah419) [inline Packs/Maltiverse/README.md:8]: There's no playbook so we can go ahead and remove this as well.

File: Packs/Maltiverse/README.md [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+## Use Cases
+---
+
```
Comment (amshamah419) [inline Packs/Maltiverse/README.md:11]: @altmannyarden Are there any use-cases we would like to put here?

File: Packs/Maltiverse/README.md [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+2. Search for Maltiverse.
+3. Click __Add instance__ to create and configure a new integration instance.
+    * __Name__: a textual name for the integration instance.
+    * __API Key__
+    * __Trust any certificate (not secure)__
+    * __Use system proxy settings__
+4. Click __Test__ to validate the URLs, token, and connection.
+## Fetched Incidents Data
+---
+
```
Comment (amshamah419) [inline Packs/Maltiverse/README.md:25]: Remove this since Maltiverse doesn't fetch incidents.

File: TestPlaybooks/playbook-Maltiverse_Test.yml [hunk 18]
```diff
@@ -0,0 +1,580 @@
+id: Maltiverse Test
+version: 11
```
Comment (amshamah419) [inline TestPlaybooks/playbook-Maltiverse_Test.yml:2]: Change this to `-1`. It's how we lock out versions and prevent version conflicts.

**Comment by amshamah419** [comment]
Since documentation isn't going to be changed, we can start the process with the tech docs team.

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py [hunk 19]
```diff
@@ -0,0 +1,422 @@
...
+        report = client.ip_report(ip)
+        positive_detections = len(report.get('blacklist', []))
+
+        blacklist_context = create_blacklist_context(report.get('blacklist', []))
+
+        outputs = {
+            'Address': report.get('ip_addr', ''),
+            'Geo.Country': report.get('country_code', ''),
+            'PositiveDetections': positive_detections,
+            'Malicious.Description': blacklist_context['Blacklist']['Description']
```
[the remark this answers is not in the document]
Comment (altmannyarden) [reply Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py:160]: Yes, I know.
I consulted with Arseny about it, and there is no option to manipulate here. So we will keep it like that.

File: Packs/Maltiverse/README.md (hunk 16 above)
Comment (altmannyarden) [reply Packs/Maltiverse/README.md:11]: Yes, like I wrote in the design:
- Enriching information about different IOC types.
- Search for Reputation about different IOC types.
- Calculate DBot Score for indicators.

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (Arsenikr) [reply Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py:160]: @altmannyarden i meant the number of levels in the son. we definitely don't want dot notation here. 
there should be two levels in the answers here for example:

`{ "geo": { "country": ""}} `

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py [hunk 20]
```diff
@@ -0,0 +1,504 @@
...
+
+            if res.status_code in [401, 403, 500]:
+                try:
+                    err_msg = str(res.json())
+                    if self.auth_token:
+                        err_msg += ' - Check server URL and API key'
+                    else:
+                        err_msg += " - Check server URL or try using an API key"
+                except ValueError:
+                    err_msg = 'Check server URL or API key'
```
Comment (yuvalbenshalom) [inline Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py:68]: if the result is not in theorem of json, you can still stringify it an add it to the error

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py [hunk 21]
```diff
@@ -0,0 +1,504 @@
...
+                    raise DemistoException('Failed to parse json object from response: {}'
+                                           .format(res.content), exception)
+
+            if res.status_code in [401, 403, 500]:
+                try:
+                    err_msg = str(res.json())
+                    if self.auth_token:
+                        err_msg += ' - Check server URL and API key'
+                    else:
+                        err_msg += " - Check server URL or try using an API key"
```
Comment (yuvalbenshalom) [inline Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.py:66]: here you are adding your line in the end of the error.
please add it in the beginning of the error.
```suggestion
                        f'Check server URL and API key \n{err_msg}'
                    else:
                        f'Check server URL or try using an API key \n{err_msg}'
```

File: Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.yml [hunk 22]
```diff
@@ -0,0 +1,421 @@
...
+      description: Qualitative maliciousness classification for an IoC. Possible values
+        are malicious, suspicious, neutral and whitelist
+      type: String
+    - contextPath: Maltiverse.File.Blacklist.FirstSeen
+      description: First time that the IoC has been seen.
+      type: Date
+    - contextPath: Maltiverse.File.Blacklist.LastSeen
+      description: Last time that the IoC was seen
+      type: Date
+    - contextPath: Maltiverse.File.Blacklist.Description
```
Comment (yuvalbenshalom) [inline Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.yml:387]: @Arsenikr / @altmannyarden are we ok with 4 level context paths?

**Comment by amshamah419** [comment]
@kirbles19 Can you take a look at the yml please? - https://github.com/demisto/content/blob/maltiverse_integration/Packs/Maltiverse/Integrations/Maltiverse/Maltiverse.yml

**Comment by kirbles19** [comment]
@amshamah419 Done.

**Pull request merged**